[Feature] Text Units

stefanwaldhauser opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Server: I added multiple integration tests (Spring) related to the features (with a high test coverage)
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/server.html).
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->

### Description
<!-- Describe your changes in detail -->
This PR activates the text-unit feature in Artemis. TextUnits can be written in Markdown (image upload also supported) and will be rendered for the student as HTML. This feature allows instructors to move some of their lecture content from the traditional lecture slides as pdf fully into Artemis

Additional Features:
- Changes to the Markdown get saved to the local storage so instructors do not lose their progress
- There is a "open as popup" feature that opens the text unit content into a separate page. This allows for easy printing and printing to PDF for students who like to have everything as PDF.



### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

Follow the videos described in the screenshot section and just test that everything works. Create, Update, Delete...

### Test Coverage
<!-- Please add the test coverage for all changes files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan -->
<!-- * ExerciseService.java: 85% -->
<!-- * programming-exercise.component.ts 95% -->
Server:
![image](https://user-images.githubusercontent.com/29718932/98488331-c4292880-2228-11eb-836a-73b3b15d368e.png)
Frontend:
![image](https://user-images.githubusercontent.com/29718932/99079068-b32d3e00-25bf-11eb-9c74-1c0d4b1a3e4d.png)
![image](https://user-images.githubusercontent.com/29718932/99079092-bc1e0f80-25bf-11eb-8a23-3f0c220df708.png)
![image](https://user-images.githubusercontent.com/29718932/99079175-dbb53800-25bf-11eb-8f46-d6bd80365163.png)


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
Create Video: https://streamable.com/ujo76r
Cache Video: https://streamable.com/m7ryv8
Student View Video: https://streamable.com/5knqvk

derLalla reviewed (changes requested):
![image](https://user-images.githubusercontent.com/33484318/98513894-e93d8b80-2268-11eb-885e-6bfa7282d999.png)

When clicking on `View in Popup` the resulting window is very small, you could try to give it a greater width and height on startup. 

Besides that this is really well done! I have tested all the other cases.

stefanwaldhauser commented:
> ![image](https://user-images.githubusercontent.com/33484318/98513894-e93d8b80-2268-11eb-885e-6bfa7282d999.png)
> 
> When clicking on `View in Popup` the resulting window is very small, you could try to give it a greater width and height on startup.
> 
> Besides that this is really well done! I have tested all the other cases.

Done in ee9efabdb8751e420a62040ef13a8a0d3ce43320

tiffca reviewed (dismissed):
Tested the test cases on test server 1 and all works well, nice changes! 

Tested creation, the cache, and viewing by student. Used sample markdown from the internet.

derLalla reviewed (dismissed):
the window size is better now. Approved.

FrankeLukas commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/textUnit.service.ts:
```diff
@@ -0,0 +1,37 @@
+import { LectureUnitService } from 'app/lecture/lecture-unit/lecture-unit-management/lectureUnit.service';
+
+type EntityResponseType = HttpResponse<TextUnit>;
+type EntityArrayResponseType = HttpResponse<TextUnit[]>;
```
GitHub Actions complains: `'EntityArrayResponseType' is defined but never used`

stefanwaldhauser replied:
Done in e881a30d6467f0e45c64b9490f1b0b6a4fef22f0

jpbernius reviewed (changes requested):
Can you check the CSS for tables? Example Markdown:
```
| Column 1       | Column 2     | Column 3     |
| :------------- | :----------: | -----------: |
|  Cell Contents | More Stuff   | And Again    |
| You Can Also   | Put Pipes In | Like this \| |
```

No borders are rendered in the box.
<img width="958" alt="image" src="https://user-images.githubusercontent.com/6382716/98666259-0a77a800-234d-11eb-952e-93076bdc3bdb.png">
It is better for the popup:
<img width="1628" alt="image" src="https://user-images.githubusercontent.com/6382716/98666293-1a8f8780-234d-11eb-9202-f3043ac580b2.png">

jpbernius commented on src/main/webapp/app/overview/course-lectures/text-unit/text-unit.component.ts:
```diff
@@ -0,0 +1,49 @@
+        const win = window.open('', ' ', 'toolbar=no,location=no,directories=no,status=no,menubar=no,scrollbars=yes,resizable=yes,' + `width=${width},height=${height}`);
+        win!.document.write(`<html><head><title>${this.textUnit.name}</title>`);
+        win!.document.write(
+            '<link rel="stylesheet" ' +
+                'href="https://cdnjs.cloudflare.com/ajax/libs/github-markdown-css/4.0.0/github-markdown.min.css" ' +
+                'integrity="sha512-Oy18vBnbSJkXTndr2n6lDMO5NN31UljR8e/ICzVPrGpSud4Gkckb8yUpqhKuUNoE+o9gAb4O/rAxxw1ojyUVzg==" crossorigin="anonymous" />',
```
Please don't embed external CS like that.

stefanwaldhauser replied:
What is the preferred way to do it?

jpbernius replied:
Best way would be to serve it from /public I guess?

stefanwaldhauser replied:
Thanks for the tip @jpbernius :) 
Done in: e2d2b5836e4fe89f36926dd66c29548072587c42

stefanwaldhauser commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/create-text-unit/create-text-unit.component.html:
```diff
@@ -0,0 +1,9 @@
+</div>
+
+<div *ngIf="!isLoading" class="container">
+    <jhi-text-unit-form [isEditMode]="false" (formSubmitted)="createTextUnit($event)"></jhi-text-unit-form>
```
Can you be more specific? What exactly should be refactored?

jpbernius replied:
I'd suggest creating a `<lecture-unit-layout/>` component with something like this as its template:
```html
<div *ngIf="isLoading" class="d-flex justify-content-center">
     <div class="spinner-border" role="status">
         <span class="sr-only">{{ 'loading' | translate }}</span>
     </div>
 </div>

 <div *ngIf="!isLoading" class="container">
     <ng-content></ng-content>
</div>
```

To be used here like so:
```html
<lecture-unit-layout>
    <jhi-text-unit-form [isEditMode]="false" (formSubmitted)="createTextUnit($event)"></jhi-text-unit-form>
</lecture-unit-layout>
```

That way the spinner and the layout would be reused for all unit types.

stefanwaldhauser replied:
Thanks for the tip. Done in 74163fbd896bbe491df078297b2024f5928bd1d4

stefanwaldhauser commented on src/main/webapp/i18n/de/lectureUnit.json:
```diff
@@ -13,6 +14,31 @@
+                "title": "Bearbeite Texteinheit"
+            },
+            "delete": {
+                "question": "Willst du wirklich die Texteinheit <strong>{{ title }}</strong> löschen?Du kannst diese Aktion nicht rückgängig machen!",
```
Done in e881a30d6467f0e45c64b9490f1b0b6a4fef22f0

stefanwaldhauser commented on src/main/webapp/i18n/de/lectureUnit.json:
```diff
@@ -13,6 +14,31 @@
+            "tooltip": "Das ist eine Texteinheit.",
+            "popup": "In Popup öffnen",
+            "notReleasedTooltip": "Text nur sichtbar für Tutoren und Dozenten. Veröffentlichkeitsdatum:",
+            "markdownHelp": "Hier kann Markdown eingegeben werden. Mehr informationen: ",
```
Done in e881a30d6467f0e45c64b9490f1b0b6a4fef22f0

stefanwaldhauser commented:
> Can you check the CSS for tables? Example Markdown:
> 
> ```
> | Column 1       | Column 2     | Column 3     |
> | :------------- | :----------: | -----------: |
> |  Cell Contents | More Stuff   | And Again    |
> | You Can Also   | Put Pipes In | Like this \| |
> ```
> 
> No borders are rendered in the box.
> <img alt="image" width="958" src="https://user-images.githubusercontent.com/6382716/98666259-0a77a800-234d-11eb-952e-93076bdc3bdb.png">
> It is better for the popup:
> <img alt="image" width="1628" src="https://user-images.githubusercontent.com/6382716/98666293-1a8f8780-234d-11eb-9202-f3043ac580b2.png">

This is broken basically everywhere where we use markdown. 
![image](https://user-images.githubusercontent.com/29718932/98686178-8b906880-2368-11eb-89de-e8ee6f13cba6.png)

stefanwaldhauser commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/edit-text-unit/edit-text-unit.component.html:
```diff
@@ -0,0 +1,9 @@
+</div>
+
+<div *ngIf="!isLoading" class="container">
+    <jhi-text-unit-form [formData]="formData" (formSubmitted)="updateTextUnit($event)" [isEditMode]="true"></jhi-text-unit-form>
```
Can you be more specific? What exactly should be refactored?

stefanwaldhauser replied:
Done in 74163fbd896bbe491df078297b2024f5928bd1d4

stefanwaldhauser commented:
> Can you check the CSS for tables? Example Markdown:
> 
> ```
> | Column 1       | Column 2     | Column 3     |
> | :------------- | :----------: | -----------: |
> |  Cell Contents | More Stuff   | And Again    |
> | You Can Also   | Put Pipes In | Like this \| |
> ```
> 
> No borders are rendered in the box.
> <img alt="image" width="958" src="https://user-images.githubusercontent.com/6382716/98666259-0a77a800-234d-11eb-952e-93076bdc3bdb.png">
> It is better for the popup:
> <img alt="image" width="1628" src="https://user-images.githubusercontent.com/6382716/98666293-1a8f8780-234d-11eb-9202-f3043ac580b2.png">

Fixed it globally in c60c2778058d5dfbe610fb8c4e40feefe19d3ff6. Now works on all markdown fields we use. The problem was that the table styling in bootstrap it opt-in, meaning the `table` class has to be added to a table for styling. I modified the markdown service to always add this class to created table tags
![image](https://user-images.githubusercontent.com/29718932/98689150-ed9e9d00-236b-11eb-9979-8f196ec50181.png)

FuchsDominik reviewed (approved):
So test wise everything seems great. Tested the features of text units and didnt experience any wrong behavior. Also the cache feature is nice.

tiffca reviewed (dismissed):
Tested successfully on test server 3. The text units with markdown work as expected, including the cache feature. Nice change!

FuchsDominik reviewed (dismissed):
Re-approve

FrankeLukas commented on src/test/javascript/spec/component/lecture-unit/text-unit/text-unit-form.component.spec.ts:
```diff
@@ -0,0 +1,182 @@
+        // mocking router
+        // mocking the local storage for cache testing
+        store = {};
+        getItemStub = sandbox.stub(localStorage, 'getItem').callsFake((key: string) => {
```
GitHub Actions: `'getItemStub' is assigned a value but never used`

stefanwaldhauser replied:
d67c60cb6977715928970ad04df3a2f288f31f6b

FrankeLukas commented on src/test/javascript/spec/component/lecture-unit/text-unit/text-unit-form.component.spec.ts:
```diff
@@ -0,0 +1,182 @@
+        getItemStub = sandbox.stub(localStorage, 'getItem').callsFake((key: string) => {
+            return store[key] || null;
+        });
+        removeItemStub = sandbox.stub(localStorage, 'removeItem').callsFake((key: string) => {
```
GitHub Actions: `'removeItemStub' is assigned a value but never used`

stefanwaldhauser replied:
d67c60cb6977715928970ad04df3a2f288f31f6b

FrankeLukas commented on src/test/javascript/spec/component/lecture-unit/text-unit/text-unit-form.component.spec.ts:
```diff
@@ -0,0 +1,182 @@
+        removeItemStub = sandbox.stub(localStorage, 'removeItem').callsFake((key: string) => {
+            delete store[key];
+        });
+        setItemStub = sandbox.stub(localStorage, 'setItem').callsFake((key: string, value: string) => {
```
GitHub Actions: `'setItemStub' is assigned a value but never used`

stefanwaldhauser replied:
done in d67c60cb6977715928970ad04df3a2f288f31f6b

FrankeLukas commented on src/test/javascript/spec/component/lecture-unit/text-unit/text-unit.component.spec.ts:
```diff
@@ -0,0 +1,117 @@
+            ],
+            providers: [
+                MockProvider(ArtemisMarkdownService, {
+                    safeHtmlForMarkdown: (markdown: string) => exampleHTML,
```
GitHub Actions: `'markdown' is defined but never used`

stefanwaldhauser replied:
i mean i need to have the parameter to mock the method? Any idea how to do this? @FrankeLukas

FrankeLukas replied:
I think it wants something like `safeHtmlForMarkdown: () => exampleHTML,` here.

stefanwaldhauser replied:
Yes but I can not do it as the mocked method has to match the signature of the original method

FrankeLukas replied:
Then I don't know. One could probably define a dummy method, but adding some sort of ignore annotation would be better if that is possible? (Unfortunately I don't know if that exists or if the errors here even matter after the PR is merged.)

stefanwaldhauser replied:
74049e8c98e6d469300d1d71e63d236bdb4a5d4d

stefanwaldhauser replied:
74049e8c98e6d469300d1d71e63d236bdb4a5d4d

FrankeLukas commented on src/test/javascript/spec/component/lecture-unit/text-unit/text-unit.component.spec.ts:
```diff
@@ -0,0 +1,117 @@
+            providers: [
+                MockProvider(ArtemisMarkdownService, {
+                    safeHtmlForMarkdown: (markdown: string) => exampleHTML,
+                    htmlForMarkdown: (markdown: string) => exampleHTML,
```
GitHub Actions: `'markdown' is defined but never used`

stefanwaldhauser replied:
i mean i need to have the parameter to mock the method? Any idea how to do this? @FrankeLukas

FrankeLukas reviewed (approved):
Followed the videos. Works as advertised.

derLalla reviewed (dismissed):
re-approve

sjagla27 reviewed (approved):
Looks good and works great. Good job :)

FuchsDominik reviewed (approved):
re-approve

derLalla reviewed (dismissed):
re-approve

nutatuna reviewed (dismissed):
Tested on my machine and it worked.

schultek reviewed (dismissed):
Creation, Editing with Cache and Deleting Text Units worked fine. For me it opened another tab instead of a popup for printing, but this will be because of my browser settings. Printing worked fine anyways. All in all very cool feature. 👍

schultek reviewed (approved):
Re-approve.

FuchsDominik reviewed (approved):
re-approve

derLalla commented on src/main/resources/public/content/github-markdown.css:
```diff
@@ -1,985 +1,735 @@
+    fill: currentColor;
+    vertical-align: text-bottom
 }
 
```
is this now an autogenerated file? Because a lot of line were deleted

stefanwaldhauser replied:
I ran it through a css cleaner

derLalla reviewed (approved):
re-approve

nutatuna reviewed (approved):
Tested the new version. Re-approve.

jpbernius commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lecture-unit-layout/lecture-unit-layout.component.ts:
```diff
@@ -0,0 +1,15 @@
+    templateUrl: './lecture-unit-layout.component.html',
+    styles: [],
+})
+export class LectureUnitLayoutComponent implements OnInit {
```
```suggestion
export class LectureUnitLayoutComponent {
```

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

jpbernius commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lecture-unit-layout/lecture-unit-layout.component.ts:
```diff
@@ -0,0 +1,15 @@
+import { Component, Input, OnInit } from '@angular/core';
```
```suggestion
import { Component, Input } from '@angular/core';
```

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

jpbernius commented on src/main/webapp/app/lecture/lecture-unit/lecture-unit-management/lecture-unit-layout/lecture-unit-layout.component.ts:
```diff
@@ -0,0 +1,15 @@
+export class LectureUnitLayoutComponent implements OnInit {
+    @Input()
+    isLoading = false;
+
+    constructor() {}
+
+    ngOnInit(): void {}
```
```suggestion
```

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

jpbernius commented on src/test/javascript/spec/component/lecture-unit/text-unit/create-text-unit.component.spec.ts:
```diff
@@ -0,0 +1,115 @@
+
+@Component({ selector: 'jhi-lecture-unit-layout', template: '<ng-content></ng-content>' })
+class LectureUnitLayoutStubComponent {
+    @Input()
+    isLoading = false;
```
```suggestion
    @Input() isLoading = false;
```

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

jpbernius commented on src/test/javascript/spec/component/lecture-unit/text-unit/edit-text-unit.component.spec.ts:
```diff
@@ -0,0 +1,157 @@
+
+@Component({ selector: 'jhi-lecture-unit-layout', template: '<ng-content></ng-content>' })
+class LectureUnitLayoutStubComponent {
+    @Input()
+    isLoading = false;
```
```suggestion
    @Input() isLoading = false;
```

stefanwaldhauser replied:
f8b3a8c32952d73bebf274e32e506db2f599a1f6

jpbernius reviewed (commented):
Some nitpicking:

stefanwaldhauser commented on src/main/resources/public/content/github-markdown.css:
```diff
@@ -0,0 +1,735 @@
+.markdown-body .octicon {
+    display: inline-block;
+    fill: currentColor;
+    vertical-align: text-bottom
```
done :)

FuchsDominik reviewed (approved):
Re-re-approve

jpbernius reviewed (approved):
Awesome! LGTM! :shipit:
